python/lsst/ts/watcher/mock_pager_duty.py

```diff
@@ -0,0 +1,239 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+__all__ = ["MockPagerDuty"]
+
+import os
+import types
+import typing
+import uuid
+from http import HTTPStatus
+
+from aiohttp import web
+
+
+def get_event_key() -> str:
+    """Return a new unique request key"""
+    return str(uuid.uuid4())
+
+
+class MockPagerDuty:
+    """A mock PagerDuty service to support Watcher escalation in unit tests.
+
+    Parameters
+    ----------
+    port : `int`
+        The TCP/IP port for the service. 0 to choose a free port.
+
+    Notes
+    -----
+    To use::
+
+        async with MockPagerDuty(port=...) as server:
+            # ... use the server
+
+        # Or if you prefer explicit start/stop:
+
+        server = MockPagerDuty(port=...)
+        await server.start()
+        # ... use the server
+        await server.close()
+
+    Attributes
+    ----------
+    port : `int`
+        If the specified port is 0, then the chosen port
+        (after `start` has run). Otherwise the specified port.
+        The default is 80 because that is what PagerDuty uses.
+    url : `str`
+        The root URL of the service. "" until `start` has run.
+    """
+
+    ALLOWED_KEYS = frozenset(
+        (
+            "payload",
+            "routing_key",
+            "event_action",
+            "dedup_key",
+            "client",
+            "client_url",
+            "links",
+            "images",
+        )
+    )
+
+    ALLOWED_PAYLOAD_KEYS = frozenset(
+        (
+            "summary",
+            "timestamp",
+            "severity",
+            "source",
+            "component",
+            "group",
+            "class",
+            "custom_details",
+        )
+    )
+
+    def __init__(self, port: int = 80) -> None:
+        self.port = port
+        self.escalation_key = os.environ["ESCALATION_KEY"]
+        # Dict of event ID: data_dict
+        self.events: dict[str, dict] = dict()
+        self.runner: web.AppRunner | None = None
+        self.site: web.TCPSite | None = None
+        self.url: str = ""
+        # Unit tests may set this True to reject the next request
+        self.reject_next_request = False
+
+    def make_app(self) -> web.Application:
+        """Make an instance of the web application."""
+        app = web.Application()
+        app.add_routes(
+            [
+                # web.post("/:{identifier}/close", self.close_event),
+                web.post("/v2/enqueue", self.send_event),
+                # web.delete("/:{identifier}", self.delete_event),
+            ]
+        )
+        return app
+
+    async def start(self) -> None:
+        """Start the service.
+
+        Raises
+        ------
+        RuntimeError
+            If port = 0 and serving on more than one socket
+            (in which case the served port cannot be determined).
+        """
```

Comment:
A RuntimeError also is raised if self.runner is not None. Please add that to the doc string.